### PR TITLE
[azservicebus] Don't clog up ReceiveMessages() with slow operations

### LINKED ISSUE #16923: [azservicebus] panic in ReceiveMessages on context timeout (invalid memory address or nil pointer dereference)
### Bug Report

- import path of package in question: ``"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"`
- SDK version (from Go.mod): 

```
        github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus v0.3.4-0.20220126013851-916bcb2ef21f
	github.com/Azure/azure-sdk-for-go/sdk/azcore v0.21.0 // indirect
	github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.3 // indirect
	github.com/Azure/azure-sdk-for-go/sdk/messaging/internal v0.0.0-20211208010914-2b10e91d237e // indirect
	github.com/Azure/go-amqp v0.17.0 // indirect
```

- output of `go version`: `go version go1.17.3 linux/amd64`

- What happened?

I call `ReceiveMessages(ctx, 95, &azservicebus.ReceiveMessagesOptions{})` and get the following panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x712de5]
goroutine 199 [running]:
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl(0xc000138150, {0x902a68, 0xc000364720}, 0x64, 0xc00038bc10)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220125201221-b3f0e1078c77/receiver.go:432 +0x265
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages(0xc000138150, {0x902a68, 0xc000364720}, 0xb1f736cc3, 0xbba120)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220125201221-b3f0e1078c77/receiver.go:184 +0x1ea
```

- How can we reproduce it?

The error seems to happen when the context times out. I tried to get a reproducible test with the code below, but it does not happen every run.

```
import (
	"context"
	"errors"
	"fmt"
	"math/rand"
	"os"
	"sync"
	"testing"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/admin"
	"github.com/stretchr/testify/require"
)

func TestAzureContextBug(t *testing.T) {
	connectionStr := os.Getenv("ASB_CONNECTION_STRING")
	require.NotEmpty(t, connectionStr, "Define ASB_CONNECTION_STRING environment variable to run this test")

	adminClient, err := admin.NewClientFromConnectionString(connectionStr, &admin.ClientOptions{})
	require.NoError(t, err)

	rand.Seed(time.Now().UnixNano())
	topicName := fmt.Sprintf("test-topic-%d", rand.Intn(1000))
	subscriptionName := "test-subscription"

	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	autoDeleteOnIdle := 5 * time.Minute
	_, err = adminClient.CreateTopic(ctx, topicName, &admin.TopicProperties{
		AutoDeleteOnIdle: &autoDeleteOnIdle,
	}, &admin.CreateTopicOptions{})
	require.NoError(t, err)

	t.Cleanup(func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		_, err := adminClient.DeleteTopic(ctx, topicName, &admin.DeleteTopicOptions{})
		require.NoError(t, err)
	})

	_, err = adminClient.CreateSubscription(ctx, topicName, subscriptionName, &admin.SubscriptionProperties{}, &admin.CreateSubscriptionOptions{})
	require.NoError(t, err)



	// publish messages


	numMessages := 1000
	batchSize := 100

	publisher := func() error {
		client, err := azservicebus.NewClientFromConnectionString(connectionStr, &azservicebus.ClientOptions{})
		if err != nil {
			return err
		}

		sender, err := client.NewSender(topicName, &azservicebus.NewSenderOptions{})
		if err != nil {
			return err
		}
		for i := 0; i < numMessages; i += batchSize {
			batch, err := sender.NewMessageBatch(ctx, &azservicebus.MessageBatchOptions{})
			if err != nil {
				return err
			}
			for j := 0; j < batchSize; j++ {
				msg := fmt.Sprintf("message-%d", i+j)
				err := batch.AddMessage(&azservicebus.Message{
					Body: []byte(msg),
				})
				if err != nil {
					return err
				}
			}
			err = sender.SendMessageBatch(ctx, batch)
			if err != nil {
				return err
			}
			//time.Sleep(10*time.Millisecond)
			t.Logf("published %d/%d messages", i+batchSize, numMessages)
		}
		t.Logf("publisher finished")
		return nil
	}
	errorChan := make(chan error)
	go func() {
		errorChan <- publisher()
	}()


	// receive messages
	client, err := azservicebus.NewClientFromConnectionString(connectionStr, &azservicebus.ClientOptions{})
	require.NoError(t, err)

	receiver, err := client.NewReceiverForSubscription(topicName, subscriptionName, &azservicebus.ReceiverOptions{})
	require.NoError(t, err)

	receivedCount := 0
	waitTime := 0 * time.Nanosecond
	for receivedCount < numMessages && ctx.Err() == nil {
		//ctx2 := newCustomContext(waitTime)
		ctx2, cancel2 := context.WithTimeout(context.Background(), waitTime)
		defer cancel2()

		t.Logf("waitTime = %v", waitTime)
		messages, err := receiver.ReceiveMessages(ctx2, 95, &azservicebus.ReceiveMessagesOptions{})
		if errors.Is(err, context.DeadlineExceeded) {
			t.Logf("context deadline exceeded")
		} else {
			require.NoError(t, err)
		}

		if len(messages) == 0 {
			continue
		}
		receivedCount += len(messages)
		t.Logf("Message count: %d (total: %d)", len(messages), receivedCount)
		waitTime = waitTime + 100*time.Nanosecond

		var wg sync.WaitGroup
		for _, message := range messages {
			body, err := message.Body()
			require.NoError(t, err)
			require.NotNil(t, body)
			//t.Logf("received %d: %s", j, body)
			wg.Add(1)
			go func(message *azservicebus.ReceivedMessage) {
				defer wg.Done()
				err = receiver.CompleteMessage(ctx, message)
				require.NoError(t, err)
			}(message)
		}
		wg.Wait()
	}
	require.Equal(t, numMessages, receivedCount)
	require.NoError(t, <- errorChan, "error in publisher")
}

```

**Edit**: Maybe also related: In other runs I got a deadlock in the same line. This is the stack trace after the test was killed after 10 minutes:

```
goroutine 1881 [select, 8 minutes]:
github.com/Azure/go-amqp.(*manualCreditor).Drain(0xc0007e8150, {0x902a30, 0xc000136000}, 0xc00038fa40)
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/manualCreditor.go:79 +0x190
github.com/Azure/go-amqp.(*link).DrainCredit(0xc00038fa40, {0x902a30, 0xc000136000})
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/link.go:607 +0x77
github.com/Azure/go-amqp.(*Receiver).DrainCredit(0x8f9ec0, {0x902a30, 0xc000136000})
	/go/pkg/mod/github.com/!azure/go-amqp@v0.17.0/receiver.go:38 +0x26
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).receiveMessagesImpl(0xc0001e8540, {0x902a68, 0xc000aac180}, 0x64, 0xc000c1d4e0)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220126013851-916bcb2ef21f/receiver.go:432 +0x282
github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus.(*Receiver).ReceiveMessages(0xc0001e8540, {0x902a68, 0xc000aac180}, 0x1, 0x11)
	/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/messaging/azservicebus@v0.3.4-0.20220126013851-916bcb2ef21f/receiver.go:184 +0x1ea
```

### LINKED ISSUE #17060: [azservicebus] Receiving duplicated messages when maxMessages is greater than 1
### Bug Report

ServiceBus receiver freezes for a minute (the message lock period) and then receives two duplicated messages (one of them with expired lock).

```
2022-02-14T10:45:46Z Receiving messages...
2022-02-14T10:46:50Z Got 2 messages...
2022-02-14T10:46:50Z Processing message: 956d72d8889e42aaa6c3e3028ddbb3ba
2022-02-14T10:46:50Z Processing message: 956d72d8889e42aaa6c3e3028ddbb3ba
2022-02-14T10:46:51Z Error completing event: *Error{Condition: com.microsoft:message-lock-lost, Description: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:c1bd8e02-2e29-43f7-b1ca-0532bb732279, TrackingId:71d9126f001701570027cc55620702b4_G10_B11, SystemTracker:securityside-development-v2:Topic:dev-contracts|facade-services-cache-updater, Timestamp:2022-02-14T10:46:51, Info: map[]}
2022-02-14T10:46:51Z Message ID: 956d72d8889e42aaa6c3e3028ddbb3ba
2022-02-14T10:46:51Z Messaged enqueued time: 2022-02-14 10:45:50.944 +0000 UTC
2022-02-14T10:46:51Z Locked until: 2022-02-14 10:46:50.959 +0000 UTC
2022-02-14T10:46:51Z Error completing event: *Error{Condition: com.microsoft:message-lock-lost, Description: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:c1bd8e02-2e29-43f7-b1ca-0532bb732279, TrackingId:71d9126f001701570027cc55620702b4_G10_B11, SystemTracker:securityside-development-v2:Topic:dev-contracts|facade-services-cache-updater, Timestamp:2022-02-14T10:46:51, Info: map[]}
2022-02-14T10:46:51Z Message ID: 956d72d8889e42aaa6c3e3028ddbb3ba
2022-02-14T10:46:51Z Messaged enqueued time: 2022-02-14 10:45:50.944 +0000 UTC
2022-02-14T10:46:51Z Locked until: 2022-02-14 10:47:50.885 +0000 UTC
```

As you can see from the above logs, the consumer froze for a minute (`2022-02-14T10:45:46Z` until `2022-02-14T10:46:50Z`), then received 2 messages with the same MessageID, one of them with the lock already expired (`2022-02-14 10:46:50.959 +0000 UTC`).

```go
func (c *ClientConsumer) receive(ctx context.Context, receiver receiver, handler consumer.EventHandler, topic string) error {
	for {
		fmt.Println("Receiving messages...")
		messages, err := receiver.ReceiveMessages(ctx, 1, nil)
		if err != nil {
			if err.Error() == "amqp: link closed" && c.closed {
				return nil
			}
			return errors.NewErrorMessage(err.Error(), false)
		}

		if len(messages) == 0 {
			continue
		}

		fmt.Printf("Got %d messages...\n", len(messages))

		wg := sync.WaitGroup{}
		wg.Add(len(messages))
		for index := range messages {
			go func(message *azservicebus.ReceivedMessage) {
				defer wg.Done()

				fmt.Printf("Processing message: %s\n", message.MessageID)
				c.handle(context.Background(), message, topic, receiver, handler)
			}(messages[index])
		}
		wg.Wait()
	}
}


// Handle handles the event.
func (c ClientConsumer) handle(ctx context.Context, e *azservicebus.ReceivedMessage, topic string, receiver receiver, handler consumer.EventHandler) {
	if e.LockedUntil != nil && e.LockedUntil.Before(time.Now()) {
		fmt.Printf("Lock already lost\nMessage ID: %s\nMessaged enqueued time: %s\nLocked until: %s\n", e.MessageID, e.EnqueuedTime, e.LockedUntil)
		return
	}

	err := handler(ctx, e)
	if err == nil {
		if err := receiver.CompleteMessage(ctx, e); err != nil {
			fmt.Printf("Error completing event: %s\nMessage ID: %s\nMessaged enqueued time: %s\nLocked until: %s\n", err.Error(), e.MessageID, e.EnqueuedTime, e.LockedUntil)
		}
		return
	}

	if err := receiver.AbandonMessage(ctx, e, nil); err != nil {
		fmt.Printf("Error abandoning event: %s\n", err.Error())
	}
}
```

Tested with: v0.3.3, v0.3.4 and v0.3.5

### PR DESCRIPTION
**Pull request by richardpark-msft**
Fixing a few bugs related to non-cancellable or long operations in ReceiveMessages():
    
* Rather than recovering in ReceiveMessages() (after messages have been received) just close the links and return. This avoids the issue in #17060 where doing the recovery inline led to us taking too long, causing us to actually get the message _again_ since the original message's lock had expired.
* Add an failsafe timeout to drain to prevent us from potentially waiting forever if the flow frame has been lost.
   
Fixes #17060
Fixes #16923 (short-circuit drain if reply doesn't come back)

### CODE REVIEW

File: sdk/messaging/azservicebus/receiver.go [hunk 1]
```diff
@@ -341,123 +353,128 @@ func (r *Receiver) receiveMessage(ctx context.Context, options *ReceiveMessagesO
...
+// or times out when draining we will close it quickly. The next operation on the link will
+// recover it.
+func flushPrefetchedMessages(ctx context.Context, receiver internal.AMQPReceiver, messages *[]*ReceivedMessage) {
+	// Draining data from the receiver's prefetched queue. This won't wait for new messages to
+	// arrive, so it'll only receive messages that arrived prior to the drain.
+	for {
+		am, err := receiver.Prefetched(ctx)
+
+		// we've removed any code of consequence from Prefetched.
+		if am == nil || err != nil || internal.IsCancelError(err) {
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/receiver.go:472]: I don't think `internal.IsCancelError(err)` clause will ever be hit when `err` isn't `nil`.

File: sdk/messaging/azservicebus/receiver.go [hunk 2]
```diff
@@ -341,123 +353,128 @@ func (r *Receiver) receiveMessage(ctx context.Context, options *ReceiveMessagesO
...
-			// d. amqplink.Receive() blocks (no messages in queue, for instance)
-			//   <years pass>
-			// e. amqpLink.Receive() returns an error
-			//
-			// We don't want to the # of attempts to already be '3' when we get to
-			// step 5, so we reset here so the next set of retry attempts starts fresh.
-			args.ResetAttempts()
-		}
-
+	err := r.amqpLinks.Retry(ctx, "receiveMessages.getlinks", func(ctx context.Context, lwid *internal.LinksWithID, args *utils.RetryFnArgs) error {
```
Comment (jhendrixMSFT) [inline sdk/messaging/azservicebus/receiver.go:360]: What exactly is being retried here?

File: sdk/messaging/azservicebus/receiver.go (hunk 1 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/receiver.go:472]: :facepalm:

File: sdk/messaging/azservicebus/receiver.go (hunk 2 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/receiver.go:360]: I have two kinds of retry "primitives". The 'non'amqpLinks' version is just a Retry() like we've got in azcore without any special integration of anything.

The amqpLinks version specifically retries on creating the connection+links and understands that specific scenario enough that, by the time it succeeds and calls your callback, it'll definitely have acquired and creating everything. So after this set of retries is complete the links should be usable.

File: sdk/messaging/azservicebus/receiver.go (hunk 1 above)
Comment (richardpark-msft) [reply sdk/messaging/azservicebus/receiver.go]: Fixed...:|

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/messaging/azservicebus/internal/amqp_test_utils.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ type FakeNS struct {
 	RPCLink         RPCLink
 	Session         AMQPSessionCloser
 	AMQPLinks       *FakeAMQPLinks
+
+	CloseCalled int
```
Comment (RickWinter) [inline sdk/messaging/azservicebus/internal/amqp_test_utils.go:22]: Nit, alignment
```suggestion
	CloseCalled       int
```

**Review by RickWinter — APPROVED** [review]

**Pull request merged**